File: proxylogcheck.pl
##################################################

# Author:      Christopher Ramirez
# Program:     Proxy Log File Analyser (PERL Script)
# Version:     1.0
# Class:       3662C Develop Software
# Instructor:  John Chetcuti
# Date:        14-10-2005

# Description:  This program will process the entries in the proxy logs 
               # generated each day and extract only the information 
               # required and present it in a suitable format.

##################################################

#------------------------
# GETTING SYSTEM TIME 
#------------------------

#_________________________________________
# Import the time functions from time module
# Get the current time
# Get the system month (full form, eg. October)
# Get the system day
# Get the system month (in short form eg. Oct)
#_________________________________________


use POSIX qw(strftime);
($Second, $Minute, $Hour, $DayOfMonth, $Month, $Year, $Weekday, $DayOfYear, $IsDST) = gmtime(time);
$Month = strftime("%B", gmtime);              
$DayOfMonth = strftime("%d", gmtime);         
$MonthDay = $Month . " " . $DayOfMonth ;       
$ShortMonth = strftime("%b", gmtime);


#-------------------------------------
# CHECKING FOR LOG FILE
#-------------------------------------

#_______________________________________________________________
# Setting the format of proxy log file as a text document
# If the proxy log file exists
    # Open the proxy file, ready for reading
# Else
    # There may be no proxy log file generated today
#_______________________________________________________________

$ProxyLog = "access " . $DayOfMonth . " " . $ShortMonth . ".txt";

if (-e $ProxyLog) {
    print "Proxy log file for ", $DayOfMonth . " " . $Month, " exists and will now be read!\n";
    print "\n";
    $OpenLog = open(PROXYLOG, "$ProxyLog");
        
}
else {
    print "No proxy log file for today."

} 


#-------------------------------------
# READING AND EXTRACTION FROM THE FILE
#-------------------------------------

#________________________________________________________________

# Read lines in the proxylog file
# sets up an output file as a list
# While the proxy log file is read
    # If "Forbidden" is found in the line
        # Create index point for start position of the username
        # Increment username length
        # Extract the username from the file
        # Increment URL start position
        # Extract the full URL from the line
#________________________________________________________________
    
$proxyline = <PROXYLOG>;         
@UsernameURLList = ();    
while ($proxyline) {             
    if ($proxyline =~ /Forbidden/i) {                      
        $IndexToUser = index($proxyline, ",", 22);         
        $Length = $IndexToUser - 22;                       
        $NaughtyUser = substr $proxyline, 22, $Length;     
        $StartURL = $IndexToUser + 13 ;                                        
        $FullURL = substr $proxyline, $StartURL;        

        
        # This section extracts a URL from the file. Some URLs differ from each other.
        # It finds the first slash, colon or space of a certain URL in the line and then extracts it
        # from the incremented URL start point to the index point.
          
        
        if (index($FullURL, "/") != -1) 
        {
            $EndURL = index($FullURL, "/");
            $URL = substr $FullURL, 0, $EndURL;
        }
        elsif (index($proxyline, ":") != -1) 
        {
            $ColonIndex = index($FullURL, ":") ;
            $URL = substr $FullURL, 0, $ColonIndex;
        } 
        elsif (index($FullURL, " ") != -1) 
        {
            $SpaceIndex = index($FullURL, " ");
            $URL = substr $FullURL, 0, $SpaceIndex;
        }

        # Concatenate strings to form a data structure to how the lines will be written to the user file
        # Create list and add the lines to the list
        # Reads the next line in the file

        $UsernameURL = $DayOfMonth ."  ". $NaughtyUser . "  " . $URL . "\n";  
        push(@UsernameURLList, $UsernameURL);                             
        $proxyline = <PROXYLOG>;                                          
    }
}

#----------------------------------------------------
# REMOVING DUPLICATE ENTRIES FROM THE LOG FILE
#----------------------------------------------------

# A hash is used to keep a record of any lines that have been seen
# Set up a list that will hold all the unique entries
# For each line in the list
    # It will then check for duplicates ignoring more than one instance of the same line
# The unique lines are added to the list
# Sort the output list into alphabetical order
# Display the output list with all the duplicates removed

%seen = ();                             
@UniqueList = ();                       
foreach $proxyline (@UsernameURLList) {
    $seen{$proxyline}++;                
}
@UniqueList = keys %seen;            
@NoDupesList = sort @UniqueList;     
print @NoDupesList;                  


#-----------------------------------------------------------
# DIRECTORY HANDLING
#-----------------------------------------------------------

# The root directory is set up
# The sub-directory in the main directory path is set up for storage of user files
           
$PathSep = "\\";
$MonthDir = $Month;
$Forbidden_Path = "C:\\Forbidden" ;                        
$ForbiddenMonthPath = $Forbidden_Path . $PathSep . $MonthDir ;   


# This checks whether the main root directory exists and if not, then it will be created.

if(-e $Forbidden_Path){
    print "--------------------------------------------------------------------------------\n";
    print "\n";
    print $Forbidden_Path, " already exists.\n";
    } 
else {
    print "--------------------------------------------------------------------------------\n";
    print "\n";
    print $Forbidden_Path, " does not exist. Constructing directory now.\n" ;   
    mkdir($Forbidden_Path);        
    }

# This checks whether the month directory exists. If not, then it will be created.

if(-e $ForbiddenMonthPath){
    print $ForbiddenMonthPath, " already exists.\n";
    } 
else {
    print $MonthDir, " directory does not exist. Constructing directory now.\n";
    mkdir($ForbiddenMonthPath);   
    }

#--------------------------------------------------------------------
# FILE HANDLING
#--------------------------------------------------------------------

# For each user who has been accessing Forbidden sites each day, a text file is generated for each user. 
# The information that is included in the text file is the date, the user and the URL that they have visited. 
# If a user has visited a site a number of times, their "crime" will only be represented as one entry per unique site. 
# A user may have accessed 5 websites which means that 5 unique entries will be added to the file.

 
# Initialise the current user
# For each line in the list
    # Create index point for username
    # Increment username length
    # Extract username from the line
    # If username is not the current user
        # Set the name of file generated under the following format
        # Set up the filepath to store the generated file
        # Generate the text file which is then ready for writing
        # Write the lines to the user files
    # Else
        # Write the Lines to the user files
    
$Current_User = "snoopdogg";                                                                                    
foreach $proxyline (@NoDupesList) {
    $IndexUser = index($proxyline, " ", 4);                              
    $Length = $IndexUser - 4;                                            
    $UsernameInLine = substr $proxyline, 4, $Length;                     
    if ($UsernameInLine ne $Current_User) {                              
        $NaughtyUserFile = $DayOfMonth . $UsernameInLine . ".txt";       
        $Filename = $ForbiddenMonthPath . $PathSep . $NaughtyUserFile;   
        $NaughtyFile = open(FILENAME, ">$Filename");                     
        $WriteInfo = print FILENAME $proxyline;                          
        $Current_User = $UsernameInLine;
    }
    else {
          $WriteInfo = print FILENAME $proxyline;
        }  
}
close(PROXYLOG);

File: proxylogcheck.py
##################################################
"""
  Author:      Christopher Ramirez
  Program:     Proxy Log File Analyser (Python Script)
  Version:     1.0
  Class:       3662C Develop Software
  Instructor:  John Chetcuti
  Date:        12-10-2005
  Description:  WWIT4 uses a program known as FreeProxy to crack down on users/students accessing
                forbidden sites that do not relate to their work or learning. A proxy log file is generated
                each day for the Administrators to view but it does not make it for easy readability.

                This script will solve this problem as it will open the proxy log file and only extract
                the information required (the username and forbidden website). This information will be
                then presented in a suitable format making it a better way for Administrators for viewing proxy
                logs.
"""
##################################################

#----------------------------------------
# GETTING SYSTEM TIME
#----------------------------------------

# Import time functions
# Get system day
# Get system month (full month)
# Get system month (short form)

from time import localtime, strftime           
Day = strftime("%d", localtime())               
Month = strftime("%B", localtime())
ShortMonth = strftime("%b", localtime())

#-----------------------------------------------
# CHECKING IF THE DAILY PROXY LOG FILE EXISTS
#-----------------------------------------------

# Importing the os and sys modules
# Setting up the name of the proxy log file to be read
# This checks whether the proxy log file exists
    # Opens up the proxy log for reading
# Else
    # There may be no proxy log file generated for the day
    # Generate an event to the event log

import os
import sys
import logging

ProxyLog = "access " + Day + " " + ShortMonth + ".txt"
print(ProxyLog)

if os.path.isfile(ProxyLog):
    print(f"Proxy Log for {Month} {Day}, exists and will now be read!")
    MY_FILE = open(ProxyLog, 'r')
else:
    from logging.handlers import NTEventLogHandler
    logger = logging.getLogger('ProxyLogCheck')
    handler = NTEventLogHandler("ProxyLogCheck")
    logger.addHandler(handler)
    logger.setLevel(logging.INFO)
    logger.info("No proxy file for today!")
    
    
#-------------------------------------------------------
# READING AND EXTRACTING INFORMATION FROM THE LOG FILE
#-------------------------------------------------------

# Read the proxy log file one line at a time
# Initialise UsernameURLList as a list

# While the proxy log file is being read
    # Initialise "Forbidden" count
    # If a Forbidden item is found in the proxy file
        # Create index point for the username
        # Initialise Naughty_User to extract username
        # These IF tests are done to find the first slash, space, colon of certain URL's
        # Initialise URL to extract URL based on the index points
        # Concatenate strings to form data structure
        # Add the string to the list
        # Read next line

proxyline = MY_FILE.readline()
UsernameURLList = []
while proxyline:                                  
    Forbiddens = proxyline.count("Forbidden")     
    if Forbiddens:                                
        IndexToUser = proxyline.find(",", 22)            
        Naughty_User = proxyline[22:IndexToUser]
        StartURL = IndexToUser + 13   
        EndURL = proxyline.find("/", StartURL)
        if EndURL != -1:
            URL = proxyline[StartURL:EndURL]
        else:
            EndURL = proxyline.find(":", StartURL)
            if EndURL != -1:
                URL = proxyline[StartURL:EndURL]
            else:
                EndURL = proxyline.find(" ", StartURL)
                URL = proxyline[StartURL:EndURL]    
                                                                                                 
        Username_URL = Day + "  " + Naughty_User + "  " + URL + "\n"     
        UsernameURLList.append(Username_URL)                             
        proxyline = MY_FILE.readline()

#_____________________________________________________________

                                   
#--------------------------------------------
# REMOVING DUPLICATE ENTRIES IN THE LOG FILE
#--------------------------------------------

# Temporarily converts the list into a dictionary and then back to a list
# Sorts the list into alphabetical order
# Prints output file with all the duplicate items removed from the list


NoDupesList = dict([(i,0) for i in UsernameURLList]).keys()
sorted(NoDupesList)
#NoDupesList.sort()                       
print(NoDupesList)


#---------------------------------------------------------
# DIRECTORY HANDLING
#---------------------------------------------------------

# Initialise path separator
# Set the main directory path to where the folders generated for each month will be stored
# Sets up the sub-directory path to where the generated user files will be stored

# If the main directory does not exist 
    # Construct the directory
    # Else, it already exists

# If the month sub-directory does not exist
    # Construct the directory
    # Else, it already exists

#____________________________________________________________
                                                
PathSep = "\\"                                           
Forbidden_Path = "C:\Forbidden"                       
ForbiddenMonthPath = Forbidden_Path + PathSep + Month    
    

if not os.path.isdir(Forbidden_Path):                        
    print
    print("Forbidden directory does not exist. Making directory now.")
    os.mkdir(Forbidden_Path)    
else:
    print
    print("Forbidden directory already exists")


if not os.path.isdir(ForbiddenMonthPath):   
    print
    print(f"{Month} directory does not exist. Making directory now.")
    os.mkdir(ForbiddenMonthPath)
else:
    print(f"{Month} directory already exists.")

#_____________________________________________________________


#-------------------------------------
# FILE HANDLING
#-------------------------------------


"""
For each user who has been accessing Forbidden sites each day, a text file is generated for that user. 
The information that is included in the text file is the date, the user and the URL that they have visited. 
If a user has visited a site a number of times, their "crime" will only be represented as one entry per unique site. 
A user may have accessed 4 websites that day which means that 4 unique entries will be added to that user file.

Example: 13 student054 www.cisco.com
         13 student054 www.hookers.com.au
         13 student054 www.chopsuey.com
         13 student054 www.mp3-mania.co.uk

"""
# Initialise the current user

# For each line in the list
    # Create index point for the user
    # Increment username length
    # Extract username from line
    # If the username value is not equal to the current user value
        # If the generated file is open
            # Close the file
        # Sets the format of the file as a text file
        # Sets the filepath as to where the generated file is to be stored
        # Generates the file, ready for writing
        # Write the lines to the user files
        # Sets the user to the current user value


#____________________________________________________________


Current_User = "snoopdogg"                    
NaughtyFile = MY_FILE
NaughtyFile.close()

for proxyline in NoDupesList:                 
    IndexUser = proxyline.find(" ", 4)        
    UsernameInLine = proxyline[4:IndexUser]   
    if UsernameInLine != Current_User:           
        if NaughtyFile is open:                        
            NaughtyFile.close()                                  
        NaughtyUserFile = Day + UsernameInLine + ".txt"                   
        Filename = ForbiddenMonthPath + PathSep + NaughtyUserFile  
        NaughtyFile = open(Filename, 'w')                                 
        WriteInfo = NaughtyFile.write(proxyline)                
        Current_User = UsernameInLine

        from logging.handlers import NTEventLogHandler
        logger = logging.getLogger("ProxyLogCheck")
        handler = NTEventLogHandler("ProxyLogCheck")
        logger.addHandler(handler)
        logger.setLevel(logging.WARNING)
        Message = "A file for", UsernameInLine, "has been created as a consequence of accessing forbidden websites." 
        logger.warning(Message)                 
    else:
        WriteInfo = NaughtyFile.write(proxyline)


   
#____________________________________________________________

MY_FILE.close()         # Close the file
    







